### PR TITLE
Perform repartitioning in compaction thread

### PR DESCRIPTION
**Pull request by knizhnik**
refer #1441

### CODE REVIEW

**Review by hlinnaka — COMMENTED** [review]
+1 on the principle.

Now that you're calling repartition() from the compact() function, this allows for some more refactoring. You can change repartition() to return the KeyPartitioning to the caller. The only caller is the compact() function, and that's also the only function that reads the LayeredTimeline.partitioning field. So you could remove LayeredTimeline.partitioning and the hint_partitioning() function altogether.

 and remove the hint_partitioning() function altogether.

**Comment by knizhnik** [comment]
> +1 on the principle.
> 
> Now that you're calling repartition() from the compact() function, this allows for some more refactoring. You can change repartition() to return the KeyPartitioning to the caller. The only caller is the compact() function, and that's also the only function that reads the LayeredTimeline.partitioning field. So you could remove LayeredTimeline.partitioning and the hint_partitioning() function altogether.

Done

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -388,6 +387,21 @@ impl<R: Repository> DatadirTimeline<R> {
 
         Ok(result.to_keyspace())
     }
+
+    pub fn repartition(&self, lsn: Lsn) -> Result<(KeyPartitioning, Lsn)> {
+        let partitioning_guard = self.partitioning.read().unwrap();
```
Comment (hlinnaka) [inline pageserver/src/pgdatadir_mapping.rs:392]: Let's change it from RwLock to Mutex. Makes it a little bit simpler. There is only this one place where the lock is used, and this is only called from the compaction thread, so there is no need to allow concurrent readers, and this isn't performance critical.

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/pgdatadir_mapping.rs (hunk 1 above)
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs]: Done

File: pageserver/src/pgdatadir_mapping.rs [hunk 2]
```diff
@@ -388,6 +387,19 @@ impl<R: Repository> DatadirTimeline<R> {
...
         Ok(result.to_keyspace())
     }
+
+    pub fn repartition(&self, lsn: Lsn) -> Result<(KeyPartitioning, Lsn)> {
+        let mut partitioning_guard = self.partitioning.lock().unwrap();
+        if partitioning_guard.1 == Lsn(0)
+            || lsn.0 - partitioning_guard.1 .0 > self.repartition_threshold
+        {
+            let keyspace = self.collect_keyspace(lsn)?;
+            let partitioning = keyspace.partition(TARGET_FILE_SIZE_BYTES);
```
Comment (lubennikovaav) [inline pageserver/src/pgdatadir_mapping.rs:397]: Hmm
Now, partitioning always uses `TARGET_FILE_SIZE_BYTES` as target_size and `TEST_FILE_SIZE` doesn't work as expected.

To enable it back, we should pass `target_size` it as a function argument.

Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs]: It seems to be time to move TARGET_FILE_SIZE_BYTES to config.
I have added `compaction_target_size` parameter.
Please propose some better name if you do not like it.

Also I want to notice that there is one more pending PR with per-tenant config which once again has to be rebased afer this change (and it is reasonable to make it possible to assign tenant-spocific `compaction_target_size`)

**Review by lubennikovaav — APPROVED** [review]

**Pull request merged**